File: js/src/native.ts
/**
 * Native / annotation tools — turn a plain function into a uniform Tool.
 * See ../../SPEC.md §6.
 */
import type { JSONSchema, Tool, ToolContext, ToolResult } from "./types.js"

export interface DefineToolOptions {
  name: string
  description: string
  inputSchema?: JSONSchema
  source?: Tool["source"]
  /** Return a string (wrapped as output) or a full ToolResult. */
  run: (args: Record<string, unknown>, ctx?: ToolContext) => unknown | Promise<unknown>
}

const EMPTY_SCHEMA: JSONSchema = { type: "object", properties: {}, additionalProperties: false }

/** Wrap a function as a Tool. The run() return value becomes the tool output. */
export function defineTool(opts: DefineToolOptions): Tool {
  return {
    name: opts.name,
    description: opts.description,
    inputSchema: opts.inputSchema ?? EMPTY_SCHEMA,
    source: opts.source ?? "native",
    async execute(args: Record<string, unknown>, ctx?: ToolContext): Promise<ToolResult> {
      try {
        const out = await opts.run(args ?? {}, ctx)
        if (out && typeof out === "object" && "output" in (out as object) && "isError" in (out as object)) {
          return out as ToolResult
        }
        return { output: typeof out === "string" ? out : JSON.stringify(out), isError: false }
      } catch (e) {
        return { output: e instanceof Error ? e.message : String(e), isError: true }
      }
    },
  }
}

/**
 * Optional decorator sugar: `@tool("description", { name, inputSchema })`.
 * Decorated methods/functions are collected; read them with `collectTools(obj)`.
 */
const REGISTRY = new WeakMap<object, Tool[]>()

export function tool(description: string, opts?: { name?: string; inputSchema?: JSONSchema }) {
  return function (value: any, context: any) {
    const fn = typeof value === "function" ? value : undefined
    if (!fn) return value
    const name = opts?.name ?? (context?.name ? String(context.name) : fn.name)
    const t = defineTool({ name, description, inputSchema: opts?.inputSchema, run: (args, ctx) => fn(args, ctx) })
    if (context?.addInitializer) {
      context.addInitializer(function (this: object) {
        const list = REGISTRY.get(this) ?? []
        list.push(t)
        REGISTRY.set(this, list)
      })
    }
    ;(fn as any).__tool = t
    return value
  }
}

/** Gather tools defined with the @tool decorator on an instance or a plain object of fns. */
export function collectTools(obj: object): Tool[] {
  const fromInstance = REGISTRY.get(obj) ?? []
  const fromProps: Tool[] = []
  for (const key of Object.keys(obj)) {
    const v = (obj as any)[key]
    if (typeof v === "function" && v.__tool) fromProps.push(v.__tool)
  }
  return [...fromInstance, ...fromProps]
}

File: golang/native.go
// Native / annotation tools — turn a plain function into a uniform Tool.
// Mirrors js/src/native.ts (defineTool) and SPEC.md §6.

package toolnexus

import (
	"context"
	"encoding/json"
	"reflect"
	"strings"
)

// emptySchema is the default input schema for a tool with no declared inputs.
func emptySchema() JSONSchema {
	return JSONSchema{
		"type":                 "object",
		"properties":           map[string]any{},
		"additionalProperties": false,
	}
}

// NativeTool wraps a plain function as a uniform Tool (source: "native").
//
// fn returning a string  ⇒ ToolResult{Output: <string>, IsError: false}.
// fn returning an error  ⇒ ToolResult{Output: err.Error(), IsError: true}.
//
// If inputSchema is nil, an empty object schema is used.
func NativeTool(
	name, description string,
	inputSchema JSONSchema,
	fn func(ctx context.Context, args map[string]any) (string, error),
) Tool {
	if inputSchema == nil {
		inputSchema = emptySchema()
	}
	return Tool{
		Name:        name,
		Description: description,
		InputSchema: inputSchema,
		Source:      SourceNative,
		Execute: func(args map[string]any, tctx *ToolContext) (ToolResult, error) {
			ctx := context.Background()
			if tctx != nil && tctx.Ctx != nil {
				ctx = tctx.Ctx
			}
			if args == nil {
				args = map[string]any{}
			}
			out, err := fn(ctx, args)
			if err != nil {
				return ToolResult{Output: err.Error(), IsError: true}, nil
			}
			return ToolResult{Output: out, IsError: false}, nil
		},
	}
}

// NativeToolReflect derives the JSON schema from struct T via `json` tags and
// builds a native Tool. The model-supplied args are decoded into T before fn is
// called.
//
// Type mapping: string→string, int/float→number, bool→boolean, slice/array→array,
// struct/map→object. A field is required unless its json tag carries ",omitempty".
// Fields tagged `json:"-"` are ignored.
func NativeToolReflect[T any](
	name, description string,
	fn func(ctx context.Context, in T) (string, error),
) Tool {
	var zero T
	schema := reflectStructSchema(reflect.TypeOf(zero))
	return Tool{
		Name:        name,
		Description: description,
		InputSchema: schema,
		Source:      SourceNative,
		Execute: func(args map[string]any, tctx *ToolContext) (ToolResult, error) {
			ctx := context.Background()
			if tctx != nil && tctx.Ctx != nil {
				ctx = tctx.Ctx
			}
			if args == nil {
				args = map[string]any{}
			}
			var in T
			// Round-trip through JSON so json tags drive the decode, matching
			// the schema we advertised.
			raw, err := json.Marshal(args)
			if err != nil {
				return ToolResult{Output: err.Error(), IsError: true}, nil
			}
			if err := json.Unmarshal(raw, &in); err != nil {
				return ToolResult{Output: err.Error(), IsError: true}, nil
			}
			out, err := fn(ctx, in)
			if err != nil {
				return ToolResult{Output: err.Error(), IsError: true}, nil
			}
			return ToolResult{Output: out, IsError: false}, nil
		},
	}
}

// reflectStructSchema builds an object JSON-Schema from a struct type.
func reflectStructSchema(t reflect.Type) JSONSchema {
	for t != nil && t.Kind() == reflect.Ptr {
		t = t.Elem()
	}
	if t == nil || t.Kind() != reflect.Struct {
		// Not a struct — fall back to a free-form object.
		return emptySchema()
	}
	props := map[string]any{}
	var required []string
	for i := 0; i < t.NumField(); i++ {
		f := t.Field(i)
		if !f.IsExported() {
			continue
		}
		name, omitempty, skip := jsonFieldName(f)
		if skip {
			continue
		}
		props[name] = reflectFieldSchema(f.Type)
		if !omitempty {
			required = append(required, name)
		}
	}
	schema := JSONSchema{
		"type":                 "object",
		"properties":           props,
		"additionalProperties": false,
	}
	if len(required) > 0 {
		schema["required"] = required
	}
	return schema
}

// jsonFieldName resolves a struct field's JSON name and omitempty flag from its
// `json` tag. skip is true for `json:"-"`.
func jsonFieldName(f reflect.StructField) (name string, omitempty, skip bool) {
	tag := f.Tag.Get("json")
	if tag == "-" {
		return "", false, true
	}
	parts := strings.Split(tag, ",")
	name = parts[0]
	if name == "" {
		name = f.Name
	}
	for _, p := range parts[1:] {
		if p == "omitempty" {
			omitempty = true
		}
	}
	return name, omitempty, false
}

// reflectFieldSchema maps a Go type to a JSON-Schema fragment.
func reflectFieldSchema(t reflect.Type) map[string]any {
	for t.Kind() == reflect.Ptr {
		t = t.Elem()
	}
	switch t.Kind() {
	case reflect.String:
		return map[string]any{"type": "string"}
	case reflect.Bool:
		return map[string]any{"type": "boolean"}
	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64,
		reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64,
		reflect.Float32, reflect.Float64:
		return map[string]any{"type": "number"}
	case reflect.Slice, reflect.Array:
		return map[string]any{
			"type":  "array",
			"items": reflectFieldSchema(t.Elem()),
		}
	case reflect.Map:
		return map[string]any{"type": "object"}
	case reflect.Struct:
		return reflectStructSchema(t)
	default:
		return map[string]any{}
	}
}
